File: app/models/abstract/categorised_dataset.rb
require 'abstract/abstract_dataset'
require 'abstract/categorised_data'
require 'pragmas'

class CategorisedDataset < AbstractDataset

  embeds_one :categorised_data, class_name: "CategorisedData", inverse_of: :categorised_dataset

  def new_data
    build_categorised_data
    categorised_data
  end
end

File: app/assets/javascripts/models/abstract/CategorisedDataset.js
define([
    'models/abstract/AbstractDataset',
    'models/abstract/CategorisedData',
], function(
    AbstractDataset,
    CategorisedData
) {

    function CategorisedDataset () {
        var _this = new AbstractDataset();

        /**
         * @returns {string}
         */
        _this.name = function() {
            return 'Categorised Dataset';
        };

        /**
         * @param {Number} year
         * @returns {CategorisedData}
         */
        _this.newData = function (year) {
            return new CategorisedData(year);
        };

        return _this;
    }
    return CategorisedDataset;
});